sdk/core/azure-core/azure/core/polling/base_polling.py

```diff
@@ -0,0 +1,550 @@
+
+    @abc.abstractmethod
+    def get_status(self, pipeline_response):
+        # type: (PipelineResponseType) -> str
+        """Return the status string extracted from this response."""
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def get_final_get_url(self, pipeline_response):
+        # type: (PipelineResponseType) -> Optional[str]
+        """If a final GET is needed, returns the URL.
+
+        :rtype: str
+        """
+        raise NotImplementedError()
+
+
+class OperationResourcePolling(LongRunningOperation):
+    """Implements a operation resource polling, typically from Operation-Location.
+
+    :param str operation_location_header: Name of the header to return operation format (default 'operation-location')
+    """
+
+    def __init__(self, operation_location_header="operation-location"):
+        self._operation_location_header = operation_location_header
+
+        # Store the initial URLs
+        self._async_url = None
+        self._location_url = None
+        self._request = None
+
+    def can_poll(self, pipeline_response):
+        """Answer if this polling method could be used.
+        """
+        response = pipeline_response.http_response
+        return self._operation_location_header in response.headers
+
+    def get_polling_url(self):
+        """Return the polling URL.
+        """
+        return self._async_url
+
+    def get_final_get_url(self, pipeline_response):
+        # type: (PipelineResponseType) -> Optional[str]
+        """If a final GET is needed, returns the URL.
+
+        :rtype: str
+        """
+        response = pipeline_response.http_response
+        if not _is_empty(response):
+            body = _as_json(response)
+            # https://github.com/microsoft/api-guidelines/blob/vNext/Guidelines.md#target-resource-location
+            resource_location = body.get("resourceLocation")
+            if resource_location:
+                return resource_location
+
+        if self._request.method in {"PUT", "PATCH"}:
+            return self._request.url
+
+        if self._request.method == "POST" and self._location_url:
+            return self._location_url
+
+        return None
+
+    def set_initial_status(self, pipeline_response):
+        # type: (PipelineResponseType) -> str
+        """Process first response after initiating long running operation.
+
+        :param azure.core.pipeline.PipelineResponse response: initial REST call response.
+        """
+        self._request = pipeline_response.http_response.request
+        response = pipeline_response.http_response
+
+        self._set_async_url_if_present(response)
+
+        if response.status_code in {200, 201, 202, 204} and self._async_url:
+            return "InProgress"
+        raise OperationFailed("Operation failed or canceled")
+
+    def _set_async_url_if_present(self, response):
+        # type: (ResponseType) -> None
+        async_url = response.headers.get(self._operation_location_header)
```

Comment:
Why do we only set `self._async_url` if there is a location header in the response? The end result is that it is `None` either way. If we want to make a stronger statement on that the initial response *must* have a operation location, then we should probably throw here. Incidentally, mypy may help determine what should happen here if we put type annotations on the `get_polling_url` method...

Reply:
fixed